### PR TITLE
[BUGFIX] Fix GDAL raster misalignement when resampling is enabled and overviews present

### LINKED ISSUE #36820: Raster Layers misaligned when resampled and uses an overview
**Describe the bug**
Is some cases resampled Geotiff images with added overviews are misaligned in the QGIS canvas by a very small amount (maybe one pixel). Changing the layer symbology resampling settings for zoomed in/out, or increasing the oversampling parameter makes no difference. 

**How to Reproduce**

1. Take a Geotiff then run the following process on it with GDAL:

```
gdal_translate source.tif out_half.tif -r cubic -outsize 50% 50%
 
cp source.tif out_with_ovr.tif
gdaladdo -r cubic out_with_ovr.tif 2
```

2. Now add both files to a QGIS project
3. Zoom to native resolution for out_half.tif
4. Toogle top layer on and off to see misaligment. 

See GIF below. You might want to open the image in a new tab to see the shift as it's only a pixel or so.

Note I've tested the output from GDAL converting the two images to a GIF with imagemagick convert and then displaying in multiple image viewers. The GDAL output looks good.

compare

**QGIS and OS versions**

QGIS master
Test on Ubuntu 18.04 and MacOS

**Additional context**

### PR DESCRIPTION
**Pull request by rouault**
Fixes #36820

### CODE REVIEW

**Comment by nyalldawson** [comment]
@rouault 

This results in completely different rendering vs the current approach.

E.g, here's a small raster (From https://github.com/qgis/QGIS/issues/30152) zoomed in, set to bilinear resampling, in current releases:

image

Here's how it looks with your change


image


The move from doing the resampling prior to applying the renderer to applying it after the renderer gives a very different result. I can see how both approaches are desirable for different use cases! There's more discussion here: https://lists.osgeo.org/pipermail/qgis-developer/2020-April/060926.html

There's also numerous tiling issues occuring here (e.g. right hand side of the above image)

image

**Comment by rouault** [comment]
Ah, I missed the mailing list discussion.

> I can see how both approaches are desirable for different use cases!

I personally prefer the new rendering :-) So what would you suggest to move forward: to have a new checkbox "Early resampling" next to the other resampling parameters ? What should be the default ?

> There's also numerous tiling issues occuring here

They are actually not tiling issues. They are due to nodata pixels and how GDAL resampling works in that case + the high zoom-in factor. Basically the resampler lacks valid data to do a good job. Using cubic resampling slightly improves things. But I don't see this as a defect.

**Comment by nyalldawson** [comment]
> I personally prefer the new rendering

I think it's totally context-dependent. E.g. the second screenshot shows quite hard, aliased edges. There's definitely going to be times when the first approach works better because it avoids these.

But for non-rendering (analysis) purposes, the first (current) approach is clearly inferior!

>  So what would you suggest to move forward: to have a new checkbox "Early resampling" next to the other resampling parameters 

Sounds good to me. And the setting should probably live in the QgsRasterPipe class (as an enum). I'd probably avoid the re-use of the existing classes here as a means of specifying the provider-side resampling, and instead add a new enum in QgsRasterDataProvider for the resampling method (then you can expose all of the gdal-supported resampling methods). If it's an explicit opt-in setting then there's no reason to restrict it to the same choices as we have for post-renderer resampling.

**Comment by rouault** [comment]
@nyalldawson I've pushed a commit hopefully implementing your suggestions.

**Comment by palmerj** [comment]
> There's definitely going to be times when the first approach works better because it avoids these.

Hi @nyalldawson thanks for looking at this. What times are you thinking about? Topo style maps with text? I assume satellite & aerial photos are going to look much better here (as well as be correctly aligned)

**Comment by nyalldawson** [comment]
@palmerj I'd say the screenshots above are a good example. The post-renderer resampling (current behavior) gives a better result IMO:
- No sharp edges, pixel boundaries are anti-aliased
- No unexpected rendering around the nodata pixels

**Comment by palmerj** [comment]
> @palmerj I'd say the screenshots above are a good example. The post-renderer resampling (current behavior) gives a better result IMO:

> No sharp edges, pixel boundaries are anti-aliased

The anti-aliased thing seems subjective. To me it looks blurry and I needs glasses when I look at it! But in general I can see AA improving some sharp edge situations - but right now that current image you have posted looks pretty bad to me. Interested in others views.

> No unexpected rendering around the nodata pixels

I wonder if the nodata issue can be fixed by @rouault. Seems like it should be possible.

Resample methods and pipelines aside, the real issue we need fixed is the misalignment of rasters on the map canvas causing spatial precision issues. Would be good to get that fixed by default for users.

**Comment by nyalldawson** [comment]
> Would be good to get that fixed by default for users.

The problem is that the new rendering is SO different from the current approach. I can't see how we could change the default setting to this without causing a lot of user confusion.

> misalignment of rasters on the map canvas causing spatial precision issues

my understanding of the linked bug report is that the difference is within 1 px -- to me this isn't a huge issue. It's an issue, yes, and would be good to ix. But honestly I would say that +/-1 pixel tolerance in rendering is completely within reasonable tolerances. I certainly can't use map tools at anything near a 1 pixel accuracy!

**Comment by rouault** [comment]
> I wonder if the nodata issue can be fixed by @rouault. Seems like it should be possible.

I'm not sure what could be changed regarding this. Resampling in GDAL (at least for RasterIO operations like here and overview computation) operates on a single band at a time (and when there's a single band, there's no alpha channel anyway). To get something smoother at the frontier between data and nodata, you would need to encode that in an alpha channel I guess (could get tricky if operating in a multi band raster where nodata pixels would not be located at the same places in different bands). So the GDAL resampler currently "just" ignore nodata pixels when convoluting source pixels contributing to a target pixel.

**Comment by nirvn** [comment]
Does provider side resampling unlocks resampling for reprojected rasters?

**Comment by palmerj** [comment]
> my understanding of the linked bug report is that the difference is within 1 px -- to me this isn't a huge issue. It's an issue, yes, and would be good to ix. But honestly I would say that +/-1 pixel tolerance in rendering is completely within reasonable tolerances. I certainly can't use map tools at anything near a 1 pixel accuracy!

I could see a difference when comparing layers, either by toggling the top layer on and off or by setting transparency on the top layer, and then this really made my question what was going on. I was really completely confused, because my data sources looked good. So I guess it's a matter of perspective. 

> I can't see how we could change the default setting to this without causing a lot of user confusion.

Well yes its a change, and definitely shouldn't be back-ported. But maybe it's a better one. That's why I wonder about testing this further with other users.

**Comment by rouault** [comment]
> Does provider side resampling unlocks resampling for reprojected rasters?

Not sure what you mean by "unlock" here. Provider side resampling occurs before QGIS raster reprojection (like current QGIS ResampleFilter)

**Comment by palmerj** [comment]
> I'm not sure what could be changed regarding this. Resampling in GDAL (at least for RasterIO operations like here and overview computation) operates on a single band at a time (and when there's a single band, there's no alpha channel anyway). To get something smoother at the frontier between data and nodata, you would need to encode that in an alpha channel I guess (could get tricky if operating in a multi band raster where nodata pixels would not be located at the same places in different bands). So the GDAL resampler currently "just" ignore nodata pixels when convoluting source pixels contributing to a target pixel.

Ok thanks Even.

**Comment by nyalldawson** [comment]
@rouault 

Here's my confusion: How does one isolated pixel end up rendering like this:

image

(it's just the bottom part of this pixel which looks wrong to me)

and yet two side-by-side looks fine:

image

**Comment by rouault** [comment]
> How does one isolated pixel end up rendering like this:

It's the pixel at -75.8758,43.2355 , right ? This is due to the palette rendering that make nodata pixels and pixels in the range 0-100 to appear as nodata. But if you look at the raw image, you'll set that above this pixel there are valid pixels in the range 0-100, whereas below it, this is pure nodata, hence the non-symetrical effect.

**Comment by rouault** [comment]
Resampling/interpolating is always "guessing" information and making some form of compromise. By doing the resampling at the provider level, one makes sure it takes into account all information about his knowledge of the georeferencing of the raster and its alignment with the exact bounds of the incoming request, as well as information about overviews. I also considered about having upper layers in the rendering pipe to be aware of that information, but I felt that would somehow mean to move the complicated resampling logic done in GDAL into QGIS. That would require major changes I think. Probably the resampling stage would have to be aware about the alignment and overviews of underlying source, and do requests to below stages that don't involve resampling, and then compensate for sub-pixel shift and resolution changes. But the results wouldn't be identical to the current resampling in the example pointed by Nyall with the color palette. Would probably be an intermediate result between current resampling and the proposed provider-level one.

I should also note that the comparison at high zoom-in factor is probably exagerating the differences in behaviour. The more you zoom in, the more you need to "invent" missing information, which will exagerate the differences between different resampling options (but all of them are somewhat arbitrary at that point. and in the above example, the current QGIS resampler operates after applying the palette, which removes some information from the source raster).

The major selling point of the provider-level resampling is primarly for zoom-out operations to fix sub pixel shifts.

**Comment by nirvn** [comment]
Testing this now. 

One nice collateral benefit here is that the resampling is _constant_ across map refresh. With the old resampling method, refreshing a map without changing the extent would result in resampled raster moving a bit, which isn't great optics to begin with.

**Comment by nirvn** [comment]
Zoomed-in cubic resampling artifacts also gone when using provider resampling:
image

@rouault , UI-wise, you'll likely need to implement a style dock equivalent to what you added in the properties dialog.

**Comment by nirvn** [comment]
Here's what I meant by shifty non-provider resampling:
Screenshot from 2020-06-12 10-20-45

Whereas provider-side resampling behaves a lot better:
Screenshot from 2020-06-12 10-20-33

_(there's a tanglible computation cost for provider-side resampling)_

**Comment by nirvn** [comment]
Biggest, win, of, all, times: resampling for reprojected rasters unlocked with a one-line patch:
Peek 2020-06-12 11-03

This is a pretty significant win, arguably eclipsing any reason to keep the post provider resampling whenever the provider can do it itself? Or at least, have that on by default.

**Comment by nirvn** [comment]
@rouault , UI wise, here's an idea: move your [x] early resampling checkbox away from symbology _into_ the rendering tab of the raster properties dialog. That way, you wouldn't need to implement it in the style dock.

**Comment by nirvn** [comment]
The line to change is in the qgsrasterprojector class, line 122 (https://github.com/qgis/QGIS/blob/master/src/core/raster/qgsrasterprojector.cpp#L122):
```
      if ( provider->capabilities() & QgsRasterDataProvider::Size )
```

Changed to:
```
      if ( !provider->isProviderResamplingEnabled() && provider->capabilities() & QgsRasterDataProvider::Size )
```

**Comment by nirvn** [comment]
Resampling on reprojected rasters would close a couple of issues  (#23602 , #23859 , and likely more not identified by my narrow search).

File: src/core/providers/gdal/qgsgdalprovider.cpp [hunk 1]
```diff
@@ -715,175 +718,296 @@ bool QgsGdalProvider::readBlock( int bandNo, int xBlock, int yBlock, void *data
...
+        if ( mZoomedOutResamplingMethod != ResamplingMethod::Nearest )
+          method = mZoomedOutResamplingMethod;
+        else
+          method = mZoomedInResamplingMethod;
+      }
+      else
+      {
+        method = mZoomedOutResamplingMethod;
+      }
+      if ( method == ResamplingMethod::Bilinear )
```
Comment (nirvn) [inline src/core/providers/gdal/qgsgdalprovider.cpp:952]: Zoomed out bilinear (i.e. the "average" combo box item) doesn't work on my machine:
Peek 2020-06-12 12-00

Comment (rouault) [reply src/core/providers/gdal/qgsgdalprovider.cpp:952]: > Zoomed out bilinear (i.e. the "average" combo box item) doesn't work on my machine:

It depends on if your image has overviews or not, and the value of the oversampling factor. With an oversampling factor of 2 and if you zoom out by more than a factor of 2 compared to the nominal resolution of the raster, no resampling will occur

Comment (nirvn) [reply src/core/providers/gdal/qgsgdalprovider.cpp:952]: That'd seem to be a regression from non provider resampling then.

This geotif I used didn't have an overview, I'd expected the zoomed out bilinear resampling with a factor of 2 to work as it used to.

Comment (nirvn) [reply src/core/providers/gdal/qgsgdalprovider.cpp:952]: @rouault , this is the geotiff I used to conduct some testing: https://storage.googleapis.com/earthenginepartners-hansen/GFC-2019-v1.7/Hansen_GFC-2019-v1.7_treecover2000_20N_100E.tif

Comment (nirvn) [reply src/core/providers/gdal/qgsgdalprovider.cpp:952]: > With an oversampling factor of 2 and if you zoom out by more than a factor of 2 compared to the nominal resolution of the raster, no resampling will occur

The QGIS logic for its zoomed out resampling is: an oversampling factor of 2 means it'll request a block twice the size and average out the values back into the requested block size. I gather that's not want GDAL does?

Comment (rouault) [reply src/core/providers/gdal/qgsgdalprovider.cpp:952]: > The QGIS logic for its zoomed out resampling is: an oversampling factor of 2 means it'll request a block twice the size and average out the values back into the requested block size. I gather that's not want GDAL does?

Yes I saw that QGIS resample filter has this logic, but GDAL doesn't do that: if you zoom out by a factor of 10 and you don't have overviews, then GDAL will need approximatively to fetch 10x10 source pixels to compute an output pixel, which will kill performance. There's no way to have accurate resampling (like GDAL does) and performance in that situation

**Comment by rouault** [comment]
> UI wise, here's an idea: move your [x] early resampling checkbox away from symbology _into_ the rendering tab of the raster properties dialog. That way, you wouldn't need to implement it in the style dock.

Hum, but would it make sense to move just that option of resampling, and not the others ? Could be confusing for people to have to switch between different tabs to control all resampling options.

File: src/core/providers/gdal/qgsgdalprovider.cpp (hunk 1 above)
Comment (nirvn) [reply src/core/providers/gdal/qgsgdalprovider.cpp:952]: Hmm, that's a significant visual regression / cost.

**Comment by nirvn** [comment]
@rouault , well, I'm thinking the early resampling is more of a one-time thing. That said, I'm starting to re-think my assessment with the discussion we're having on zoomed out resampling.

**Comment by nirvn** [comment]
So, my opinion after more testing:
- zoomed in resampling, I can see a fair amount of objective reasons why we'd want provider-side  on  by default (resampling on reprojected datasets being the obvious reason, consistency across renderings while panning, no-data artifacts gone). How it looks at the end is a bit more subjective, I find the provider-side resampling superior on the datasets I tried.
- zoomed out resampling, that's a different issue; QGIS offered a nice logic which while imperfect worked really well at making rasters look smoother. We'd have ppl complaining about the visual regression.

Understanding the GDAL logic on how it deals with zoomed out oversampling, I can get things to look good again, except that I have to push the oversampling ratio to 15+ to cover an average country size in Southeast Asia.

**Comment by nirvn** [comment]
GDAL, avg. zoomed out sampling, oversampling factor of 16:
Screenshot from 2020-06-12 18-16-06

QGIS, avg. zoomed out sampling, oversampling factor of 2:
Screenshot from 2020-06-12 18-16-11

**Comment by rouault** [comment]
>     * We'd have ppl complaining about the visual regression.

If that'd be useful, I can imagine that provider-side resampling could try to fallback to using QgsRasterResampleFilter when there's not enough overviews with respect to the max oversampling factor, instead of just falling back to GDAL nearest neighbour resampling. The consequence of course is that subpixel shifts would reappear when this code path is triggered

> The line to change is in the qgsrasterprojector class, line 122 (https://github.com/qgis/QGIS/blob/master/src/core/raster/qgsrasterprojector.cpp#L122):

My intended scope of work doesn't include fixing reprojection. While I see your suggestion can improve things and I wouldn't mind applying it, a proper fix would be to do resampling in QgsRasterProjector (like gdalwarp does) instead at the data acquisition stage. Actually one would probably want to do reprojection just after data acquisition in the pipe, instead at at the end of it like currently. With some difficult compromise when reprojecting AND downsampling AND overviews not present...

**Comment by nirvn** [comment]
@rouault , a fallback code path would be good.

On the reprojected resampling, if the one line change provides acceptable results and can be a temporary solution until a day the right approach is coded, I'd say go for it. I'm not intimate enough with the raster code to make that final judgement.

**Comment by rouault** [comment]
@nirvn With my latest commits, I think I've adressed your points

**Comment by nirvn** [comment]
Fantastic. UI/UX wise, this is IMHO ready. 

There's one more logic that would be worth adding here: if the raster is being reprojected, always use provider-side resampling if available (that's better than not working at all).

**Comment by rouault** [comment]
> There's one more logic that would be worth adding here: if the raster is being reprojected, always use provider-side resampling if available (that's better than not working at all).

Done

**Comment by nirvn** [comment]
@rouault , thanks, looking great.

Is the plan to merge this onto 3.14 prior to its release? If so I'd say the sooner the better.

**Comment by nyalldawson** [comment]
I'd like to review before merge please

**Comment by palmerj** [comment]
I would like to review too. Haven't had time yet.

**Comment by nirvn** [comment]
Here's another benefit from provider side resampling: the contour renderer doesn't pixelate when zooming in.

File: python/core/auto_generated/raster/qgsrasterdataprovider.sip.in [hunk 2]
```diff
@@ -517,6 +518,96 @@ does not support this functionality), an empty point is returned.
...
+    ResamplingMethod zoomedOutResamplingMethod() const;
+%Docstring
+Returns resampling method for zoomed-out operations.
+
+.. versionadded:: 3.16
+%End
+
+    virtual bool setMaxOversampling( double factor );
+%Docstring
+Set rmaximum oversampling factor forzoomed-out operations.
```
Comment (nyalldawson) [inline python/core/auto_generated/raster/qgsrasterdataprovider.sip.in:591]:
```suggestion
Set maximum oversampling factor for zoomed-out operations.
```

File: python/core/auto_generated/raster/qgsrasterdataprovider.sip.in [hunk 3]
```diff
@@ -517,6 +518,96 @@ does not support this functionality), an empty point is returned.
...
+
+:return: true if success
+
+.. versionadded:: 3.16
+%End
+
+    double maxOversampling() const;
+%Docstring
+Returns maximum oversampling factor forzoomed-out operations.
```
Comment (nyalldawson) [inline python/core/auto_generated/raster/qgsrasterdataprovider.sip.in:600]:
```suggestion
Returns maximum oversampling factor for zoomed-out operations.
```

File: src/core/providers/gdal/qgsgdalprovider.cpp [hunk 4]
```diff
@@ -44,6 +44,10 @@
 #include "qgssettings.h"
 #include "qgsogrutils.h"
 
+#include "qgsrasterresamplefilter.h"
+#include "qgsbilinearrasterresampler.h"
+#include "qgscubicrasterresampler.h"
```
Comment (nyalldawson) [inline src/core/providers/gdal/qgsgdalprovider.cpp:49]: No longer required?

File: src/core/providers/gdal/qgsgdalprovider.cpp [hunk 5]
```diff
@@ -715,175 +718,388 @@ bool QgsGdalProvider::readBlock( int bandNo, int xBlock, int yBlock, void *data
...
   }
 
-  // srcBottom must be less than raster height or we'll get a raster I/O error,
-  // this may happen because of rounding errors with the floating point operations used above
-  // See: issue GH #34435
-  srcBottom = std::min( mHeight - 1, srcBottom );
+  const int srcWidth = srcRight - srcLeft + 1;
+  const int srcHeight = srcBottom - srcTop + 1;
+
+  // Use GDAL cresampling if asked and possible
```
Comment (nyalldawson) [inline src/core/providers/gdal/qgsgdalprovider.cpp:864]:
```suggestion
  // Use GDAL resampling if asked and possible
```

File: src/core/providers/gdal/qgsgdalprovider.cpp [hunk 6]
```diff
@@ -715,175 +718,388 @@ bool QgsGdalProvider::readBlock( int bandNo, int xBlock, int yBlock, void *data
...
+        if ( mZoomedOutResamplingMethod != ResamplingMethod::Nearest )
+          method = mZoomedOutResamplingMethod;
+        else
+          method = mZoomedInResamplingMethod;
+      }
+      else
+      {
+        method = mZoomedOutResamplingMethod;
+      }
+      if ( method == ResamplingMethod::Bilinear )
```
Comment (nyalldawson) [inline src/core/providers/gdal/qgsgdalprovider.cpp:950]: switch

File: src/core/providers/gdal/qgsgdalprovider.cpp [hunk 7]
```diff
@@ -715,175 +718,388 @@ bool QgsGdalProvider::readBlock( int bandNo, int xBlock, int yBlock, void *data
...
+                          << QStringLiteral( "PIXELOFFSET=%1" ).arg( dataSize )
+                          << QStringLiteral( "LINEOFFSET=%1" ).arg( dataSize * tmpWidth )
+                          << QStringLiteral( "DATAPOINTER=%1" ).arg( reinterpret_cast< qulonglong >( tmpBlock ) ) );
+    GDALAddBand( hSrcDS.get(),  GDALGetRasterDataType( gdalBand ), papszOptions );
+    CSLDestroy( papszOptions );
+
+    GDALRasterIOExtraArg sExtraArg;
+    INIT_RASTERIO_EXTRA_ARG( sExtraArg );
+
+    if ( mZoomedOutResamplingMethod == ResamplingMethod::Bilinear )
```
Comment (nyalldawson) [inline src/core/providers/gdal/qgsgdalprovider.cpp:1049]: switch

File: src/core/raster/qgsrasterinterface.h [hunk 8]
```diff
@@ -31,6 +31,8 @@
 #include "qgsrasterhistogram.h"
 #include "qgsrectangle.h"
 
+class QgsRasterResampleFilter;
```
Comment (nyalldawson) [inline src/core/raster/qgsrasterinterface.h:34]: is this needed?

File: src/gui/raster/qgsresamplingutils.cpp [hunk 9]
```diff
@@ -0,0 +1,186 @@
...
+                                      QDoubleSpinBox *maximumOversamplingSpinBox,
+                                      QCheckBox *cbEarlyResampling )
+{
+  mRasterLayer = rasterLayer;
+  mZoomedInResamplingComboBox = zoomedInResamplingComboBox;
+  mZoomedOutResamplingComboBox = zoomedOutResamplingComboBox;
+  mMaximumOversamplingSpinBox = maximumOversamplingSpinBox;
+  mCbEarlyResampling = cbEarlyResampling;
+
+  mZoomedInResamplingComboBox->insertItem( 0, QObject::tr( "Nearest neighbour" ) );
```
Comment (nyalldawson) [inline src/gui/raster/qgsresamplingutils.cpp:45]: The nicer way to handle this kind of thing is to set the user data on the item to the enum value, and then use that to retrieve /set the combobox value. See .e.g 

https://github.com/qgis/QGIS/blob/master/src/gui/layout/qgslayouthtmlwidget.cpp#L62
https://github.com/qgis/QGIS/blob/master/src/gui/layout/qgslayouthtmlwidget.cpp#L182
https://github.com/qgis/QGIS/blob/master/src/gui/layout/qgslayouthtmlwidget.cpp#L435

File: python/core/auto_generated/raster/qgsrasterdataprovider.sip.in (hunk 2 above)
Comment (rouault) [reply python/core/auto_generated/raster/qgsrasterdataprovider.sip.in:591]: fixed

File: python/core/auto_generated/raster/qgsrasterdataprovider.sip.in (hunk 3 above)
Comment (rouault) [reply python/core/auto_generated/raster/qgsrasterdataprovider.sip.in:600]: fixed

File: src/core/providers/gdal/qgsgdalprovider.cpp (hunk 4 above)
Comment (rouault) [reply src/core/providers/gdal/qgsgdalprovider.cpp:49]: fixed

File: src/core/providers/gdal/qgsgdalprovider.cpp (hunk 5 above)
Comment (rouault) [reply src/core/providers/gdal/qgsgdalprovider.cpp]: fixed

File: src/core/providers/gdal/qgsgdalprovider.cpp (hunk 6 above)
Comment (rouault) [reply src/core/providers/gdal/qgsgdalprovider.cpp]: fixed

File: src/core/providers/gdal/qgsgdalprovider.cpp (hunk 7 above)
Comment (rouault) [reply src/core/providers/gdal/qgsgdalprovider.cpp]: fixed

File: src/core/raster/qgsrasterinterface.h (hunk 8 above)
Comment (rouault) [reply src/core/raster/qgsrasterinterface.h]: fixed

File: src/gui/raster/qgsresamplingutils.cpp (hunk 9 above)
Comment (rouault) [reply src/gui/raster/qgsresamplingutils.cpp]: fixed

**Comment by rouault** [comment]
I've added the option for Cubic resampling for zoomed out too (for both "classic" and provider resampling)

**Comment by nirvn** [comment]
I would really like to see this land prior to 3.14. Are we there yet?

**Comment by palmerj** [comment]
I've done testing with our topo maps and aerial photo rasters and the quality is good. The raster misalignement is fixed when using bilinear or cubic resampling. Thanks @rouault!

**Comment by rouault** [comment]
With my latest commit, I believe I'm done

**Comment by nyalldawson** [comment]
I'm happy with the code, but I think we should defer till post 3.14.0. There's just not enough time left before the 3.14.0 release for this to be tested sufficiently

**Comment by palmerj** [comment]
> tested sufficiently

What test is required? Manual user testing?

**Comment by nyalldawson** [comment]
> What test is required? Manual user testing?

Yes, on a wide variety of systems/data/etc

It's not a new regression, so we really can't rush this fix in. (I get that it's important for you, but stability of the release is a much more important decision)

**Comment by palmerj** [comment]
> It's not a new regression, so we really can't rush this fix in. (I get that it's important for you, but stability of the release is a much more important decision)

It's not that important, just good to fix. Was wondering more about the process. It seems we should be trying add as many automatic tests as possible with the required data, so it's not left to QC too much by the community.

**Comment by nyalldawson** [comment]
>  It seems we should be trying add as many automatic tests as possible with the required data, so it's not left to QC too much by the community.

Yes - totally! But there's always going to be unplanned situations/unexpected data which we just can't predict in advance...

**Comment by jscarto** [comment]
It is fantastic to see new resampling options added. However, could the previous 'average' resampling be reverted? The loss of this option is really unfortunate, and bilinear and cubic are nice additions but should not be a replacement. Average is the best approach in many cases, and its loss is a major detriment.

 https://twitter.com/jscarto/status/1341764315070693377

**Comment by rouault** [comment]
> The loss of this option is really unfortunate

It wasn't lost. What was called "average" in QGIS UI was in fact calling GDAL bilinear resampling.

**Comment by jscarto** [comment]
> `What was called "average" in QGIS UI was in fact calling GDAL bilinear resampling.

I see the text was changed, but at some point the functionality has, too. The 'average' resampling, with oversampling control, used to work exactly as expected. The latest bilinear resampling produces a much different result.

resampling

**Comment by rouault** [comment]
> I see the text was changed, but at some point the functionality has, too.

you could file an enhancement request to add AVERAGE resampling

**Comment by jscarto** [comment]
Thanks! I've just done that (a bit new to OS feature requests etc, so I appreciate the help/tips).

https://github.com/qgis/QGIS/issues/40746

**Pull request merged**